cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+optionally including the statement S to be signed.
+
+2. Commitment: Each node i picks a random secret vi and
+computes its individual commit Vi = Gvi . In a bottom-up
+process, each node i waits for an aggregate commit Vˆj from
+each immediate child j, if any. Node i then computes its
+own aggregate commit Vˆi = Vi \prod{j ∈ Cj}{Vˆj}, where Ci is the
+set of i’s immediate children. Finally, i passes Vi up to its
+parent, unless i is the leader (node 0).
+
+3. Challenge: The leader computes a collective challenge
+c = H( Aggregate Commit ∥ Aggregate Public key || Message ),
+then multicasts c down through the tree, along
+with the statement S to be signed if it was not already
+announced in phase 1.
+
+4. Response: In a final bottom-up phase, each node i waits
+to receive a partial aggregate response rˆj from each of
+its immediate children j ∈ Ci. Node i now computes its
+individual response ri = vi + cxi, and its partial aggregate
+response rˆi = ri + \sum{j ∈ Cj}{rˆj} . Node i finally passes rˆi
+up to its parent, unless i is the root.
+*/
+package cosi
+
+import (
+	cryptoRand "crypto/rand"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/dedis/crypto/abstract"
+	//own "github.com/nikkolasg/learning/crypto/util"
+)
+
+// Cosi is the struct that implements the basic cosi.
+type Cosi struct {
+	// Suite used
+	suite abstract.Suite
+	// publics is the list of public keys used for signing
+	publics []abstract.Point
+	// mask is the mask used to select which signers participated in this round
+	// or not. All code regarding the mask is directly inspired from
+	// github.com/bford/golang-x-crypto/ed25519/cosi code.
+	mask *CosiMask
+	// the longterm private key we use during the rounds
+	private abstract.Secret
+	// the message being co-signed
+	message []byte
+	// timestamp of when the announcement is done (i.e. timestamp of the four
+	// phases)
+	timestamp int64
+	// random is our own secret that we wish to commit during the commitment phase.
+	random abstract.Secret
+	// commitment is our own commitment
+	commitment abstract.Point
+	// V_hat is the aggregated commit (our own + the children's)
+	aggregateCommitment abstract.Point
+	// globalComitment is the global commit of all participants
+	//passed down in the challenge phase
+	globalCommitment abstract.Point
+	// challenge holds the challenge for this round
+	challenge abstract.Secret
+	// response is our own computed response
+	response abstract.Secret
+	// aggregateResponses is the aggregated response from the children + our own
+	aggregateResponse abstract.Secret
+}
+
+// NewCosi returns a new Cosi struct given the suite, the longterm secret, and
+// the list of public keys. If some signers were not to be participating, you
+// have to set the mask using `SetMask` method. By default, all participants are
+// designated as participating.
+func NewCosi(suite abstract.Suite, private abstract.Secret, publics []abstract.Point) *Cosi {
```

Comment:
@ineiti :
> CoSi shouldn't care how this mask is generated, anyway, this is the protocol's or service's task.

Well, not really in my opinion because CoSi *is* the one generating the mask and the mask *is part of* its protocol... But it does not change anything for CoSi which cosigners are missing and so on.
Actually, in the comments it says, 
```
// If some signers were not to be participating, you
// have to set the mask using `SetMask` method. By default, all participants are
// designated as participating.
```
Then I forgot to implement the `SetMask` method , which I'm doing right now. I find this solution more elegant than having to specify a mask explicitely *if you don't need it*.
(1)
```
// No missing cosigners
c := NewCosi(suite,secret,NewCosiMask(suite,publics))
...
// Missing cosigners
mask := NewCosiMask(suite,publics)
mask.SetBitMask(4,false)
c := NewCosi(suite,secret,mask)
```
vs (2)
```
// No missing cosigners
cosi := NewCosi(suite,secret,publics)
...
// Missing cosigners
mask := NewCosiMask(suite,publics)
mask.SetBitMask(4,false)
c := NewCosi(suite,secret,publics)
c.SetMask(mask)
``` 
Or (2- bis)
```
// No missing cosigners
cosi := NewCosi(suite,secret,publics)
...
// Missing cosigners
mask := NewCosiMask(suite,publics)
mask.SetBitMask(4,false)
c := NewCosiWithMask(suite,secret,mask)
```

One approach is shorter for the standard behavior (2), and the other one is reducing the code for the exception behavior (1), but if you use (2-bis), then it's the same amount. I would go for the (2-bis) solution, but as democracy imposes us a majority, let the majority decide :)